File: Lintcode 018 - Subsets 2.java
Lintcode 018 - Subsets 2.java

class Solution {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsetsWithDup(ArrayList<Integer> nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>> ();
        ArrayList<Integer> list = new ArrayList<Integer>();
        
        if(nums == null || nums.size() == 0) {
            return result;
        }
        
        Collections.sort(nums);
        dfs(result, list, nums, 0);
        return result;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> result, ArrayList<Integer> list, ArrayList<Integer> nums, int pos) {
        result.add(new ArrayList<Integer>(list));
        
        for(int i = pos; i < nums.size(); i++) {
            if(i != pos && nums.get(i) == nums.get(i - 1)) {
                continue;
            }
            list.add(nums.get(i));
            dfs(result, list, nums, i + 1);
            list.remove(list.size() - 1);
        }
    }
}

File: Lintcode 018 - Subsets 2.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] nums = new int[3]{1,1,4};
            List<List<int>> result = new List<List<int>>{};   
            List<int> list = new List<int>{};
            // bool[] used = new bool[4];

            Subset(nums, result, list, 0);

            foreach(var clist in result)
            {
                foreach(var num in clist)
                {
                    Console.Write(num + ", ");
                }
                Console.WriteLine(" ");
            }
        }

        static void Subset(int[] nums, List<List<int>> result, List<int> list, int pos)
        {

            result.Add(new List<int>(list));
            for(int i = pos; i< nums.Length; i++)
            {
                if(i == pos ||  nums[i - 1] != nums[i])
                {
                    list.Add(nums[i]);
                    Subset(nums, result, list, i + 1);
                    list.Remove(list[list.Count - 1]);
                }
            }
        }
    }
}
